lib/events/pgevents/pgevents.go

```diff
@@ -0,0 +1,584 @@
+	}
+
+	l.log.Info("Started events backend.")
+
+	return l, nil
+}
+
+// Log is an external [events.AuditLogger] backed by a PostgreSQL database.
+type Log struct {
+	log  logrus.FieldLogger
+	pool *pgxpool.Pool
+
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+// Close stops all background tasks and closes the connection pool.
+func (l *Log) Close() error {
+	l.cancel()
+	l.wg.Wait()
+	l.pool.Close()
+	return nil
+}
+
+var schemas = []string{
+	`CREATE TABLE events (
+		event_time timestamptz NOT NULL,
+		event_id uuid NOT NULL,
+		event_type text NOT NULL,
+		session_id uuid NOT NULL,
+		event_data json NOT NULL,
+		creation_time timestamptz NOT NULL DEFAULT now(),
+		CONSTRAINT events_pkey PRIMARY KEY (event_time, event_id)
+	);
+	CREATE INDEX events_creation_time_idx ON events USING brin (creation_time);
+	CREATE INDEX events_search_session_events_idx ON events (session_id, event_time, event_id)
+		WHERE session_id != '00000000-0000-0000-0000-000000000000';`,
+}
+
+// setupAndMigrate sets up the database schema, applying migrations in order if
+// necessary.
+func (l *Log) setupAndMigrate(ctx context.Context) error {
+	var version int32
+	var migrateErr error
+
+	if err := pgcommon.Retry0(ctx, l.log, func() error {
+		_, err := l.pool.Exec(ctx,
+			`CREATE TABLE IF NOT EXISTS audit_version (
+				version integer PRIMARY KEY CHECK (version > 0),
+				created timestamptz NOT NULL DEFAULT now()
+			)`, pgx.QueryExecModeExec,
+		)
+		return trace.Wrap(err)
+	}); err != nil {
+		// the very first SELECT in the next transaction will fail, we don't
+		// need anything higher than debug here
+		l.log.WithError(err).Debug("Failed to confirm the existence of the audit_version table.")
+	}
+
+	if err := pgcommon.RetryTx(ctx, l.log, l.pool, pgx.TxOptions{
+		IsoLevel:   pgx.Serializable,
+		AccessMode: pgx.ReadWrite,
+	}, func(tx pgx.Tx) error {
+		if err := tx.QueryRow(ctx,
+			"SELECT COALESCE(max(version), 0) FROM audit_version",
+			pgx.QueryExecModeExec,
+		).Scan(&version); err != nil {
+			return trace.Wrap(err)
+		}
+
+		if int(version) > len(schemas) {
+			migrateErr = trace.BadParameter("unsupported schema version %v", version)
+			return nil
+		}
+
+		if int(version) == len(schemas) {
+			return nil
+		}
+
+		for _, s := range schemas[version:] {
+			if _, err := tx.Exec(ctx, s, pgx.QueryExecModeExec); err != nil {
+				return trace.Wrap(err)
+			}
+		}
+
+		if _, err := tx.Exec(ctx,
+			"INSERT INTO audit_version (version) VALUES ($1)",
+			pgx.QueryExecModeExec, len(schemas),
+		); err != nil {
+			return trace.Wrap(err)
+		}
+
+		return nil
+	}); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if migrateErr != nil {
+		return migrateErr
+	}
+
+	if int(version) != len(schemas) {
+		l.log.WithFields(logrus.Fields{
+			"previous_version": version,
+			"current_version":  len(schemas),
+		}).Info("Migrated database schema.")
+	}
+
+	return nil
+}
+
+// periodicCleanup removes events past their retention period from the table,
```

Comment:
```suggestion
// periodicCleanup removes events past the retention period from the table,
```

`their` suggests that the retention period is per item, which is not the case.